### ISSUE TITLE
Collection class should be optional in relation mapping

### ISSUE THREAD

**Issue by leongersen** [phpstan/phpstan-doctrine#400]
I'm working with an application that does not rely on Doctrine's collections in its entities.

I have an entity like this:
```PHP
class Foo {
    /** @var iterable<Bar> */
    #[ORM\OneToMany(mappedBy: 'foo', targetEntity: Bar::class)]
    private iterable $bars = [];
}
```

PhpStan reports:
```
Property Domain\Foo::$bars type mapping mismatch: property can contain iterable<Domain\Bar>
    but database expects Doctrine\Common\Collections\Collection&iterable<Domain\Bar>.
```

While `iterable<Domain\Bar>` is technically broader than `Doctrine\Common\Collections\Collection&iterable<Domain\Bar>`, the types are compatible. Doctrine does not require relations in properties to be collections, and also supports arrays.

I expected no error to be reported here.

**Comment by ondrejmirtes** [comment]
Hi, I think your assumption is wrong. If the property contains some other iterable type than Collection, Doctrine isn't going to be able to persist it.

**Comment by leongersen** [comment]
I guess I should be more specific; it's not that any `iterable` is supported, but `array` specifically is. 

However, I can't typehint to `array`, because Doctrine will hydrate relations with `Collection`s.

**Comment by ondrejmirtes** [comment]
Please find something about this in the documentation, I couldn't: https://www.doctrine-project.org/projects/doctrine-orm/en/2.13/reference/association-mapping.html

All entities I've ever seen were using Collection in this scenario.

**Comment by leongersen** [comment]
Ok, this is fun. I can't find an explicit mention in the documentation, but it is supported behavior in both ORM 2.x and 3.0. Looking at the git blame before that gets a little messy as it concerns commits of 14+ years ago, but it looks like this used to be the default way of doing things in ORM 1.

References:
- [Documentation for model designs with arrays for relations in ORM 1](https://www.doctrine-project.org/projects/doctrine1/en/latest/manual/defining-models.html#one-to-many-and-many-to-one
);
- [The code design switches from using array internally to using collections internally](https://github.com/doctrine/orm/commit/2ec4cc5cbe7a4fdc58cca00c89e154d773b6669a#diff-55a900494fc8033ab498c53929716caf0aa39d6bdd7058e7d256787a24412ee4);
- [Support for arrays as relations in ORM 2](https://github.com/doctrine/orm/blob/2.13.x/lib/Doctrine/ORM/UnitOfWork.php#L682);
- [That same support in ORM 3](https://github.com/doctrine/orm/blob/3.0.x/lib/Doctrine/ORM/UnitOfWork.php#L576);

To conclude, relations do not need to be collections, Doctrine ORM will convert arrays to ArrayCollection where needed.

**Comment by ondrejmirtes** [comment]
I don't think that the `iterable` typehint is really valuable. You can only do it to pass data to `foreach`. You can't add anything into the collection, you can't ask about a specific element (like `->first()`).

Since this isn't really documented on the Doctrine side, I don't think you should rely on it. Thanks for understanding.

**Comment by leongersen** [comment]
Yes, I understand your opinion as this isn't documented for ORM. Thank you for your attention today, you're doing great work!
